keras_nlp/models/falcon/falcon_attention.py

```diff
@@ -0,0 +1,236 @@
+    """FalconAttention layer used in Falcon models.
+
+    This layer implements the attention mechanism used in Falcon models.
+    It performs multi-head self-attention on the input hidden states.
+
+    Args:
+        config: Configuration object containing the hyperparameters.
+
+    Input Shapes:
+        - hidden_states: Tensor of shape `(batch_size, seq_length, hidden_size)`.
+        - alibi: Tensor of shape `(batch_size, num_heads, 1, kv_length)` containing
+          alibi values for each head. Set to None if not using alibi.
+        - attention_mask: Tensor of shape `(batch_size, 1, 1, seq_length)` containing
+          the attention mask. Set to None if not using attention mask.
+        - layer_past: Tuple of tensors `(past_key, past_value)` containing the cached
+          key and value states from previous layers. Set to None if not using cache.
+        - head_mask: Tensor of shape `(num_heads,)` containing the head mask. Set to None
+          if not using head mask.
+        - use_cache: Boolean value indicating whether to use the cache.
+        - output_attentions: Boolean value indicating whether to output attention scores.
+
+    Output Shapes:
+        - output_tensor: Tensor of shape `(batch_size, seq_length, hidden_size)`.
+        - present: Tuple of tensors `(key_layer, value_layer)` containing the updated
+          key and value states for caching.
+        - attention_probs (optional): Tensor of shape `(batch_size, num_heads, seq_length, seq_length)`
+          containing the attention scores. Only present if `output_attentions=True`.
+
+    Examples:
+        ```python
+        config = AttentionConfig(hidden_size=768, n_head=12, head_dim=64, hidden_dropout=0.1,
+                                 bias=True, multi_query=False, attention_dropout=0.1)
+        attention = FalconAttention(config)
+        output_tensor, present = attention(hidden_states, alibi, attention_mask, layer_past=None,
+                                           head_mask=None, use_cache=False, output_attentions=False)
+        ```
+    """
+
+    def __init__(self, config):
+        super(FalconAttention, self).__init__()
+
+        self.hidden_size = config.hidden_size
+        self.num_heads = config.n_head
+        self.head_dim = self.hidden_size // self.num_heads
+        self.split_size = self.hidden_size
+        self.hidden_dropout = config.hidden_dropout
+
+        if self.head_dim * self.num_heads != self.hidden_size:
+            raise ValueError(
+                f"`hidden_size` must be divisible by num_heads (got `hidden_size`: {self.hidden_size} and `num_heads`:"
+                f" {self.num_heads})."
+            )
+
+        self.maybe_rotary = FalconRotaryPositionalEmbedding(config.head_dim) if config.rotary else lambda q, k: (q, k)
+
+        # Layer-wise attention scaling
+        self.inv_norm_factor = 1.0 / math.sqrt(self.head_dim)
+        self.beta = self.inv_norm_factor
+
+        self.query_key_value = Dense(
```

Comment:
We can probably split this layer into three seperate EinsumDense layers. Can you verify?